src/libostree/ostree-kernel-args.c

```diff
@@ -81,6 +81,255 @@ _ostree_kernel_args_free (OstreeKernelArgs *kargs)
   g_free (kargs);
 }
 
+void
+_ostree_kernel_args_cleanup (void *loc)
+{
+  _ostree_kernel_args_free (*((OstreeKernelArgs**)loc));
+}
+
+
+static gboolean
+strcmp0_equal (gconstpointer v1,
+               gconstpointer v2)
+{
+  return g_strcmp0 (v1, v2) == 0;
+}
+
+/* Note: this function is newly added to the API */
+GHashTable*
+_ostree_kernel_arg_get_kargs_table (OstreeKernelArgs *kargs)
+{
+
+  if (kargs != NULL)
+    return kargs->table;
+  return NULL;
+}
+
+/* Note: this function is newly added to the API */
+GPtrArray*
+_ostree_kernel_arg_get_key_array (OstreeKernelArgs *kargs)
+{
+  if (kargs != NULL)
+    return kargs->order;
+  return NULL;
+}
+
+/**
+ * _g_ptr_array_find_with_equal_func: (skip)
+ * @haystack: pointer array to be searched
+ * @needle: pointer to look for
+ * @equal_func: (nullable): the function to call for each element, which should
+ *    return %TRUE when the desired element is found; or %NULL to use pointer
+ *    equality
+ * @index_: (optional) (out caller-allocates): return location for the index of
+ *    the element, if found
+ *
+ * Checks whether @needle exists in @haystack, using the given @equal_func.
+ * If the element is found, %TRUE is returned and the element’s index is
+ * returned in @index_ (if non-%NULL). Otherwise, %FALSE is returned and @index_
+ * is undefined. If @needle exists multiple times in @haystack, the index of
+ * the first instance is returned.
+ *
+ * @equal_func is called with the element from the array as its first parameter,
+ * and @needle as its second parameter. If @equal_func is %NULL, pointer
+ * equality is used.
+ *
+ * Returns: %TRUE if @needle is one of the elements of @haystack
+ * Since: 2.54
+ */
+gboolean
+_g_ptr_array_find_with_equal_func (GPtrArray     *haystack,
```

Comment:
Just a few points:

I like that this is prefixed with `_` to indicate it's local. I think going a step further, we could prefix it with `ot` instead to indicate it's defined within OSTree (i.e. `ot_ptr_array_find_with_equal_func`).

I think this is OK using the implementation from GLib - however could you add a note to the comment above that this is copied in order to use the function from a later version of GLib without making OSTree depend on it, and a link to the GLib source code?

I think a good place for this function definition to live would be in `src/libotutil/ot-tool-util.c` (ostree keeps some private, but reusable (within ostree's codebase) in `src/libotutil`), and add the function definition to the header to `src/libotutil/ot-tool-util.c`. Then, other code outside of `ostree-kernel-args.c` can use this if needed.

Just to note when a similar situation occurred (as @jlebon noted earlier) in rpm-ostree, a slightly different implementation was used.
https://github.com/projectatomic/rpm-ostree/blob/5ca9c731d70aca44c092a6720e8a38fb0bd059d7/src/libpriv/rpmostree-kargs-process.c#L117. However, it should be OK copying the GLib function as long as there is a link to the original source.

Reply:
Thanks you for the review! I'm renaming the function to `ot_ptr_array_find_with_equal_func` and noted the source url. The function now is moved to `src/libotutil/ot-tool-util.c`. I think I might leave it the same as GLib function for now since if we are using newer version of GLib then minimum adjustment needs to be made. Will be pushing and testing the changes tomorrow. Thank you again for the feedback!